src/native/corehost/hostmisc/trace.cpp

```diff
@@ -13,9 +13,34 @@
 //  COREHOST_TRACE=1 COREHOST_TRACE_VERBOSITY=1          implies g_trace_verbosity = 1.  // Trace "enabled".  error() messages will be produced
 static int g_trace_verbosity = 0;
 static FILE * g_trace_file = stderr;
-static pal::mutex_t g_trace_mutex;
 thread_local static trace::error_writer_fn g_error_writer = nullptr;
 
+namespace
+{
+    class spin_lock
+    {
+    public:
+        spin_lock() = default;
+        spin_lock(const spin_lock&) = delete;
+        spin_lock& operator=(const spin_lock&) = delete;
+
+        void lock()
+        {
+            while (flag.test_and_set(std::memory_order_acquire)) { /*spin*/ }
```

Comment:
You might want to yield at some point. Perhaps spin 1024 times or something and then yield? This is also optional if there is low concern about this code path.

```suggestion
            uint32_t spin = 0;
            while (flag.test_and_set(std::memory_order_acquire))
            { 
                if (spin++ % 1024 == 0)
                    std::this_thread::yield();
            }
```